feat: add base catalog concept page (#359)
flox/floxdocs: docs/concepts/base-catalog.md

Reviewer: Overarching comments:
- You need to add this page to the nav, which is defined in `mkdocs.yaml`.
- Please put each sentence on a new line, it makes adding review comments/suggestions easier.

---
title: What is the Base Catalog?
description: Everything you need to know about the Base Catalog
---

# What is the Base Catalog

The base Catalog is the root of all dependencies in Flox.  It is generated from

> a fork of [nixpkgs][nixpkgs], contains a set of historical versions and is

Reviewer: ```suggestion
a fork of [nixpkgs][nixpkgs], contains historical metadata (such as versions and system support over time), and is
```

> updated regularly.  When you install something with Flox, it will come from or

Reviewer: ```suggestion
updated on an automated schedule.  When you install something with Flox, it will come from or
```

I want it to be clear that this isn't an ad-hoc, manual process.

will be built upon the Base Catalog data.

> On a regular basis, Flox will evaluate a subset of packages from it's

Reviewer: ```suggestion
On a regular basis, Flox will evaluate a subset of packages from its
```

[nixpkgs][nixpkgs] fork and save the metadata from the evaluation.  This makes
it available to users of Flox.

## What packages does Flox evaluate

Generally, Flox evaluates `legacyPackages.<system>.*` and follows the
`recurseForDerivations` attribute.  In addition to this, some explicit paths are

> evaluated.  The following is the current set, but is likely to change on a

Reviewer: ```suggestion
evaluated.  The following is the current set, but is likely to change over time.
```

regular basis.

For all system types:

- `nodePackages`
- `rustPlatform.rustLibSrc`
- `nodePackages_latest`

Additionally for `darwin` system types:

- `darwin`
- `swiftPackages`
- `swiftPackages.darwin`

## How often dose this happen

tldr; Flox ingests the equivalent of nixos-unstable on a daily basis.  So you
should expect Flox package availability to follow nixos-unstable by, at most,

> about a day.

Reviewer: Try this instead:

> Evaluating all of nixpkgs on every commit of the nixpkgs repository is computationally expensive and often unnecessary.
> Instead, Flox ingests the equivalent of the `nixos-unstable` branch on a daily basis.
> This means that once a package has landed on `nixos-unstable` you can expect it in the Base Catalog within a day.
> 
> It's important to note that this is distinct from the time between merging a pull request on the nixpkgs repository and when that package becomes available in the Base Catalog.
> The nixpkgs repository performs a series of checks after merging a pull request, and changes are merged into a series of different branches as different checks are performed on the attributes (packages) changed by the pull request.
> This process can take anywhere from a couple of days to over a week depending on whether the change causes breakage in other packages in the nixpkgs repository.
> 
> If you have submitted a pull request to nixpkgs and are interested in tracking its progress, you can use this site: [Nixpkgs PR Tracker](https://nixpk.gs/pr-tracker.html).

exposed in the UX.  These are not to be confused with upstream branches of the
same name, and all derive from the `unstable` branch of our fork[floxpkgs].

- `unstable` is reset to upstream `unstable` daily
- `staging` is reset to the `unstable` branch of our [fork][floxpkgs] each Saturday
- `stable` is reset to the `staging` branch of our [fork][floxpkgs] on the first Saturday of the month
- `lts` is reset to the `stable` branch of our [fork][floxpkgs] on the first Saturday in January and July (every 6 months)

These "stability" channels will be exposed in future versions allowing users to
select a varying frequency of updates.  This architecture also allows for
patches and backporting of fixes against all stabilities.

Note that these branches do NOT reflect upstream branches of similar name, or
release channels.  Backporting that occurs on those branches is not yet
available via Flox.

### Delays

Since the Flox Base Catalog is based on upstream nixos-unstable, it is subject
to any delays that occur during the regular operations of the nix eco-system.
If there is a breakage or if an update causes significant re-builds, it may be

> deferred to a longer process and further delay it's arrival on nixos-unstable,

Reviewer: ```suggestion
deferred to a longer process and further delay its arrival on `nixos-unstable`,
```

regular basis.

For all system types:

- `nodePackages`
- `rustPlatform.rustLibSrc`
- `nodePackages_latest`

Additionally for `darwin` system types:

- `darwin`
- `swiftPackages`
- `swiftPackages.darwin`

## How often dose this happen

tldr; Flox ingests the equivalent of nixos-unstable on a daily basis.  So you
should expect Flox package availability to follow nixos-unstable by, at most,
about a day.

### Flox branches

Flox maintains a fork of [nixpkgs][floxpkgs] and maintains several branches.
These branches relate to a Flox concept of "stabilities" that is not yet fully
exposed in the UX.  These are not to be confused with upstream branches of the
same name, and all derive from the `unstable` branch of our fork[floxpkgs].

- `unstable` is reset to upstream `unstable` daily
- `staging` is reset to the `unstable` branch of our fork[floxpkgs] each Sunday
- `stable` is reset to the `staging` branch of our fork[floxpkgs] monthly
- `lts` is reset to the `stable` branch of our fork[floxpkgs] every 6 months

These "stability" channels will be exposed in future versions allowing users to
select a varying frequency of updates.  This architecture also allows for

> patches and backporting of fixes against all stabilities.

Reviewer 2: Good callout!
